password_enhancements.module

```diff
@@ -0,0 +1,403 @@
+      '#descriptions' => $descriptions,
+      '#hidden' => empty($form['account']['pass']['#required']),
+      '#weight' => $form['account']['pass']['#weight'] + 0.001,
+      '#attached' => [
+        'library' => $libraries,
+        'drupalSettings' => [
+          'passwordEnhancementsConstraint' => [
+            'updateEffect' => Drupal::configFactory()->get('password_enhancements.settings')->get('constraint_update_effect') ?? Settings::CONSTRAINT_EFFECT_STRIKETHROUGH,
+            'plugins' => $plugin_config,
+            'minimumRequiredConstraints' => $policy->getMinimumRequiredConstraints(),
+          ],
+        ],
+      ],
+      '#prefix' => "<div id='{$ajax_wrapper}'>",
+      '#suffix' => '</div>',
+    ];
+
+    if (!empty($form['account']['roles'])) {
+      $form['account']['roles']['#ajax'] = [
+        'callback' => 'password_enhancements_update_constraints_by_roles',
+        'event' => 'change',
+        'wrapper' => $ajax_wrapper,
+      ];
+    }
+  }
+
+  $form['#validate'][] = 'password_enhancements_form_user_register_form_alter_validate';
+}
+
+/**
+ * Validate callback for password_enhancements_form_user_register_form_alter().
+ *
+ * @param array $form
+ *   The form.
+ * @param \Drupal\Core\Form\FormStateInterface $form_state
+ *   The current form state.
+ *
+ * @throws \Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException
+ * @throws \Drupal\Component\Plugin\Exception\PluginNotFoundException
+ * @throws \Drupal\Component\Plugin\Exception\PluginException
+ */
+function password_enhancements_form_user_register_form_alter_validate(array &$form, FormStateInterface $form_state): void {
+  if (!empty($password = $form_state->getValue('pass'))) {
+    $roles = $form_state->getValue('roles') ?? [];
+
+    /** @var \Drupal\password_enhancements\Entity\Storage\PolicyEntityStorageInterface $password_policy_storage */
+    $password_policy_storage = Drupal::entityTypeManager()->getStorage('password_enhancements_policy');
+    $policy = $password_policy_storage->loadByRoleAndPriority($roles);
+
+    $constraints = _password_enhancements_get_constraints($roles);
+    $error_messages = [];
+    $minimum_required_constraints = $policy->getMinimumRequiredConstraints();
+    $valid_count = 0;
+    foreach ($constraints as $constraint) {
+      /** @var \Drupal\password_enhancements\Plugin\PasswordConstraintPluginInterface $plugin_instance */
+      $plugin_instance = Drupal::service('plugin.manager.password_constraint')->createInstance($constraint->getType(), $constraint->getConfiguration());
+
+      if (!$plugin_instance->validate($password)) {
+        $error_messages[] = $plugin_instance->getErrorMessage();
+      }
+      else {
+        $valid_count++;
+      }
+
+      if ($minimum_required_constraints <= $valid_count) {
+        break;
+      }
+    }
+
+    // Display any validation error messages.
+    if ($error_messages !== [] && $valid_count < $minimum_required_constraints) {
```

Comment:
`if ($error_messages)` or `if(empty($error_messages)` would not be enough? :D Why do you need to compare it with an empty array like this? This is not JavaScript :) Is it faster?

Reply:
Two reasons, I don't like automatic typecasting and calling empty() on this set wouldn't make sense, because I know the exact type of this variable, it cannot be anything else.